### DISCUSSION TITLE
Confusion about initialization procedure.

### DISCUSSION THREAD

**Discussion by mindrunner** [fastify/fastify#5177] [Q&A] [answered]
Today, I tried to implement the `notFoundHandler` for the `rate-limit` plugin mentioned here: https://github.com/fastify/fastify-rate-limit#preventing-guessing-of-urls-through-404s

However, it turned out that `fastify.rateLimit` was `undefined` when doing
```
  fastify.setNotFoundHandler(
    {
      preValidation: fastify.rateLimit(),
    },
    notFoundHandler
  )
```

After adding an `await fastify.after()` to wait for initialization, this seemed to have done the trick. However, my custom `errorHandler` stopped working:

```
  fastify.register(....)
  fastify.register(....)
  fastify.register(....)

  await fastify.after()

  fastify.setErrorHandler(errorHandler)

  fastify.setNotFoundHandler(
    {
      preValidation: fastify.rateLimit(),
    },
    notFoundHandler
  )

```

Swapping `setErrorHandler` with the await of `after` re-enables it.

```
  fastify.register(....)
  fastify.register(....)
  fastify.register(....)

  fastify.setErrorHandler(errorHandler)

  await fastify.after()

  fastify.setNotFoundHandler(
    {
      preValidation: fastify.rateLimit(),
    },
    notFoundHandler
  )

```

Can someone explain me what's happening here. Also, how am I supposed to do it the 'correct' way?

**Comment by metcoder95** [comment, marked as answer]
Hey! I can shade some light here:

The `onRoute` hook is synchronous, meaning that it will take immediate effect after being called. On the other hand, `plugins` register happens asynchronously.

`fastify-rate-limit` depends on `onRoute` hook to set up the app for it. As you need it as a dependency of another step, the plugin hasn't been resolved yet when you are attempting to use it immediately after.

To solve it, you need to wait for the plugin to be fully registered before moving to the next step.

You can achieve that by waiting for the after (which means to chain plugin dependencies), or just calling ` await fastify.register(plugin)` and waiting for it before moving forward.

TL;DR
- `onRoute` is synchronous
- Plugin register is not

**Reply by mindrunner** [reply]
Perfectly makes sense, thank you :)

**Reply by leandroandrade** [reply]
To contribute to this answer, another way to do this is using `dependencies` of plugins. The repo below shows this approach with `@fastify/autoload`. Works really well.

https://github.com/leandroandrade/fastify-cjs
